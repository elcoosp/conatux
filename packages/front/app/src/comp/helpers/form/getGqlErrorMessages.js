const getGqlErrorMessages = error =>
  error === undefined ? [] : error.graphQLErrors.map(({ message }) => message)

export default getGqlErrorMessages
