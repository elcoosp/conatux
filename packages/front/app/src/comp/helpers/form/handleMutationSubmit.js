/* namedMutationObj: An object with a unique property which is the name of the mutation and whose value is a mutation function */
const execNamedMutation = async (namedMutationObj, variables) => {
  const [mutationName] = Object.keys(namedMutationObj)
  const mutation = namedMutationObj[mutationName]

  const {
    data: { [mutationName]: response } // Extract the relevant response data from an apollo result
  } = await mutation({
    variables
  })

  return response
}

// Handle submitting an apollo mutation from a formik onSubmit event
const handleMutationSubmit = namedMutationObj => async (
  variables, // Mutation variables
  { setSubmitting, setErrors } // Formik handler function
) => {
  try {
    const response = await execNamedMutation(namedMutationObj, variables)
    console.log(response)
  } catch (error) {
    console.log(error.graphQLErrors)
  } finally {
    setSubmitting(false)
  }
}

export default handleMutationSubmit
