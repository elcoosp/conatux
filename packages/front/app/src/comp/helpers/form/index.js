export { default as LabelledField } from './LabelledField'
export { default as Form } from './Form'
