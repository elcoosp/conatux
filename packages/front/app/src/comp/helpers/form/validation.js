import * as yup from 'yup'

const messages = {
  required: fieldName => `${fieldName} is required`,
  valid: fieldName => `${fieldName} must be a valid one`
}

const username = yup
  .string()
  .min(4)
  .max(12)
  .required()
  .default('')

const email = yup
  .string()
  .email(messages.valid('Email'))
  .max(40)
  .required(messages.required('Email'))
  .default('')

const password = yup
  .string()
  .min(12)
  .max(100)
  .required(messages.required('Password'))
  .default('')

const passwordConfirmation = yup
  .string()
  .oneOf([yup.ref('password'), null], 'Passwords must match') // null is mandatory here (cf. https://github.com/jaredpalmer/formik/issues/90)
  .required(messages.required('Password confirmation'))
  .default('')

export const fields = {
  email,
  username,
  password,
  passwordConfirmation
}

const validation = {
  makeSchemaFor: (...fieldNames) => {
    return yup.object().shape(
      fieldNames.reduce((acc, name) => {
        acc[name] = fields[name]
        return acc
      }, {})
    )
  }
}

export default validation
