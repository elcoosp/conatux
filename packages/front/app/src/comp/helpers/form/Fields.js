import React from 'react'
import { LabelledField } from '@:/comp/helpers/form'

export const Username = () => (
  <LabelledField label="Username" type="text" name="username" />
)

export const Email = () => (
  <LabelledField label="Email" type="email" name="email" />
)

export const Password = () => (
  <LabelledField label="Password" type="password" name="password" />
)

export const PasswordConfirmation = () => (
  <LabelledField
    label="Password confirmation"
    type="password"
    name="passwordConfirmation"
  />
)
export default {
  Username,
  Email,
  Password,
  PasswordConfirmation
}
