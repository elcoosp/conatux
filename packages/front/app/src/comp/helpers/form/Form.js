import React from 'react'
import P from 'prop-types'
import { Formik, Form as FormikForm } from 'formik'
import { Form as UiForm, Submit } from 'ui/form'

const FormInner = UiForm.withComponent(FormikForm)

const Form = ({ children, ...props }) => (
  <Formik {...props}>
    {({ isSubmitting, errors }) => (
      <FormInner>
        {children}
        <Submit disabled={isSubmitting || Object.keys(errors).length > 0} />
      </FormInner>
    )}
  </Formik>
)

Form.propTypes = {
  children: P.node.isRequired
}

export default Form
