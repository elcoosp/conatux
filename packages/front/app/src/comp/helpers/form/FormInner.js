import React from 'react'

import P from 'prop-types'
import { Form as UiForm, Submit } from 'ui/form'

const FormInner = ({ children, isSubmitting, handleReset, handleSubmit }) => (
  <UiForm onReset={handleReset} onSubmit={handleSubmit}>
    {children}
    <Submit disabled={isSubmitting} />
  </UiForm>
)

FormInner.propTypes = {
  children: P.node.isRequired,
  isSubmitting: P.bool.isRequired,
  handleReset: P.func.isRequired,
  handleSubmit: P.func.isRequired
}

export default FormInner
