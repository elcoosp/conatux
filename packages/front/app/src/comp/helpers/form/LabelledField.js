import React from 'react'
import P from 'prop-types'
import { Label, Input } from 'ui/form'
import { Field as FormikField, ErrorMessage } from 'formik'

const Field = Input.withComponent(FormikField)
const LabelledField = ({ label, name, type }) => (
  <>
    <Label>
      {label}
      <Field name={name} type={type} />
    </Label>
    <ErrorMessage name={name} />
  </>
)

LabelledField.propTypes = {
  label: P.string.isRequired,
  name: P.string.isRequired,
  type: P.string.isRequired
}
export default LabelledField
