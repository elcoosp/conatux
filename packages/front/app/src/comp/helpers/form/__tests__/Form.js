import React from 'react'
import { shallow } from 'enzyme'
import { Submit } from 'ui/form'
import { Formik } from 'formik'

import Errors from '@:/comp/helpers/Errors'

import { desc, getFakeProps, expected } from '@:/utils/testing'
import Form, { isDisabled, FormInner } from '../Form'

const [props] = getFakeProps(__dirname, '../Form')('Form')

desc.compWithSnap({ Form }, props, () => {
  describe('Mutation handler', () => {
    it('should pass an `onSubmit` prop to the `Formik` component if the `mutation` prop is an object', () => {
      const wrapper = shallow(<Form {...props} mutation={{}} />)
      expect(
        wrapper
          .find(Formik)
          .first()
          .prop('onSubmit')
      ).toBeDefined()
    })

    it('should not pass an `onSubmit` prop to the `Formik` component if the `mutation` prop is falsy', () => {
      const wrapper = shallow(<Form {...props} mutation={undefined} />)
      expect(
        wrapper
          .find(Formik)
          .first()
          .prop('onSubmit')
      ).toBe(undefined)
    })
  })
  describe('Form children', () => {
    const setup = ({ isSubmitting = false, errors = {} } = {}) =>
      shallow(<Form {...props} />).renderProp('children')({
        isSubmitting,
        errors
      })

    expected.components(setup(), { Submit, FormInner, Errors })
  })
})

desc.module('isDisabled', () => {
  it('should return true if isSubmitting is true', () => {
    expect(isDisabled(true)).toBe(true)
  })

  it('should return true if errors has some keys', () => {
    expect(isDisabled(false, { foo: 'bar' })).toBe(true)
  })

  it('should return false if both isSubmitting is false and errors has no keys', () => {
    expect(isDisabled(false, {})).toBe(false)
  })
})
