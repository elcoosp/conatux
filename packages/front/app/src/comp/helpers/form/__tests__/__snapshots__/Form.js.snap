// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should match snapshot 1`] = `
<Formik
  enableReinitialize={false}
  isInitialValid={false}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={
    Object {
      "cast": [Function],
    }
  }
>
  <Component />
</Formik>
`;
